scenarios/22_North_Pole.cfg

```diff
@@ -23,6 +23,85 @@
         [/lua]
     [/event]
 
+    #define IMPROVE_STATS_ON_FROZEN UNIT_FILTER
+        #arg DEF
+        60
+        #endarg
+
+        [object]
+            id=frozen_bonus
+            duration=scenario
+            silent=yes
+            take_only_once=no
+            [filter]
+                {UNIT_FILTER}
+            [/filter]
+            [effect]
+                apply_to=defense
+                replace=yes
+                [defense]
+                    frozen={DEF}
+                [/defense]
+            [/effect]
+            [effect]
+                apply_to=movement_costs
+                replace=yes
+                [movement_costs]
+                    frozen=2
+                [/movement_costs]
+            [/effect]
+        [/object]
+    #enddef
+
+    #make summoned ground troops move faster on snow
+    [event]
+        name=unit placed,post advance,post metamorph
+        first_time_only=no
+        [filter]
+            side=1,4,5
+        [/filter]
+
+        {VARIABLE unit_type_when_bonus_set $unit.variables.unit_type_when_bonus_set}
+
+        {IF_VAR unit_type_when_bonus_set not_equals $unit.type (
+            # Skipped if cached.
+            [then]
+                # Cache is cleaned on reloads, so always remove object first to avoid duplication.
+                [remove_object]
+                    id=$unit.id
+                    object_id=frozen_bonus
+                [/remove_object]
+
+                # Note that Fire Elemental doesn't fly.
+                {IF_UNIT_INTERSECT id=$unit.id (
+                    type=EoMa_Carpet_Rider,EoMa_Carpet_Master,EoMa_Dimensional_Gate,EoMa_Dimensional_Gate_II,EoMa_Jinn,EoMa_Great_Jinn,EoMa_Wonderful_Jinn,EoMa_Mystical_Jinn,EoMa_Efreet,EoMa_Great_Efreeti,EoMa_Air_Elemental,EoMa_Air_Avatar,EoMa_Air_God,EoMa_Fire_Avatar,EoMa_Fire_God
+                ) (
+                    # flying
+                ) (
+                    # grounded
+                    {VARIABLE def_bonus 60}
+                    {IF_VAR unit.type equals EoMa_Um (
+                        [then]
+                            {VARIABLE def_bonus 70}
+                        [/then]
+                    )}
+                    {IMPROVE_STATS_ON_FROZEN (id=$unit.id) DEF=$def_bonus}
+                    {CLEAR_VARIABLE def_bonus}
+                )}
+                # Cache unit type ("unit placed" is spammed on [harm_unit], and who knows elsewhen.)
+                {MODIFY_UNIT id=$unit.id variables.unit_type_when_bonus_set $unit.type}
+            [/then]
+        )}
+        {CLEAR_VARIABLE unit_type_when_bonus_set}
+
+        [allow_undo]
+        [/allow_undo]
+        [on_undo]
+            # invalidate cache
+            {MODIFY_UNIT id=$unit.id unit.variables.unit_type_when_bonus_set $empty}
```

Comment:
why this is needed?

Reply:
Just for completeness. Unless I am missing something, it shouldn't be good to change the unit state on an action, but keep the changes when undoing it.